hwbench/environment/block_devices.py

```diff
@@ -0,0 +1,100 @@
+import pathlib
+from json import loads
+
+import pyudev
+
+from hwbench.utils.external import External
+
+
+class Block_Device:
+    """Block_Device is a class that gathers block_device SMART data using smartctl"""
+
+    def __init__(self, out_dir, udev_device):
+        self.udev_device = udev_device
+        self.name = udev_device.sys_name
+        self.type = self.get_block_device_type()
+        self.out_dir = out_dir
+        self.smart_json = b"{}"
+
+    def is_rotational(self) -> bool:
+        syspath = f"/sys/block/{self.udev_device.sys_name}/queue/rotational"
+        with open(syspath) as file:
+            content = file.read()
+        return bool(content.strip())
+
+    def is_ata(self) -> bool:
+        bus = self.udev_device.get("ID_BUS")
+        return bool(bus == "ata")
+
+    def is_nvme(self) -> bool:
+        return bool(self.udev_device.parent.get("NVME_TRTYPE"))
+
+    def get_block_device_type(self):
+        if self.is_ata() and self.is_rotational():
+            return "hdd"
+        if not self.is_rotational():
+            return "ssd"
+
+    def get_smart(self):
+        """Dump SMART information for all block devices"""
+        smart_info = Smartctl(self.out_dir, self.name)
+        return smart_info.dump()
+
+
+class Block_Devices:
+    """Block_Devices is a class that gets a list of block_devices using pyudev and holds a collection of corresponding Block_device objects"""
+
+    udev_context = pyudev.Context()
+    data: dict[str, Block_Device] = {}
+
+    def __init__(self, out_dir: pathlib.Path):
+        self.out_dir = out_dir
+        self.discover_devices()
+
+    def discover_devices(self):
```

Comment:
Can users call this function ?
If not, it should be named __discover_devices()
If so, then you should probably reset self.data to avoid ghost data

Reply:
I don't see why they could not. I could imagine some testing that would reconfigure disks namespaces/partitions. retriggering could be something useful. wdyt ?

Reply:
I'm fine if they call it but here the case that creates ghost data:

System has 1 disk, sda.
I instanciate Block_devices() which call discover_devices()
self.data has 1 disk : sda

I'm adding a new disk(sdb) and call discover_devices()
self.data has two disks, sda & sdb

I'm removing sda and call discover_devices()
self.data still has two disks, sda & sdb

I think your discover_devices function should ensure the data structure is empty at every call unless you'll have ghost data.

Reply:
okay. made private method